### ISSUE TITLE
Training data used for pre-trained model.

### ISSUE THREAD

**Issue by bhardwajvijay** [sampepose/flownet2-tf#22]
Hello,

Very nice work, and thanks for the code. I just wanted to know about the datasets being used for training the pre-trained model. 

or particularly I wanted to know that if you used KITTI training databases during training the model??

Thanks and Regards,
Vijay

**Comment by yinjunbo** [comment]
@bhardwajvijay 
Hello~
I just have the same confusion.
Did you get an answer?
Thank you~

**Comment by bhardwajvijay** [comment]
Hi @yinjunbo,

Yeah, I have this information. It is first trained on Chairs dataset and then finetuned on the Things3D dataset. Author did not used KITTI for training the model.

Regards,
Vijay

**Comment by yinjunbo** [comment]
@bhardwajvijay 
Thank you for your reply.
If it's convenient I want to know whether you test the trained model on the Sintel or KITTI dataset.
I found that using the model provided by the author leads to an AEE about 5.8 with flownet_c or
flownet_s.
Is there anything I should pay attention to?

Best regards~
Junbo

**Comment by bhardwajvijay** [comment]
Hi Junbo,

I tested only flownet2 version on KITTI training database (for which I had ground truth available). I got following numbers:

**KITTI 2012:** 
                 Average End Point Error (AEPE): 2.18
                 Outlier Percentage (Flow vectors with AEPE greater than 3): 13.7 %
**KITTI 2015:** 
                 Average End Point Error (AEPE): 5.80
                 Outlier Percentage (Flow vectors with AEPE greater than 3): 26.3 %

KITTI 2015 has higher AEPE because of the fact that it contains dynamic objects in the scene, which were not present in KITTI 2012 database.

Hope it helps. 

Regards,
Vijay

**Comment by yinjunbo** [comment]
@bhardwajvijay , Thank you very much for your helpful reply !
According your result, the model seems to work well on KITTI dataset.
Many thanks for your kind message.

All the best.
Junbo

**Comment by bhardwajvijay** [comment]
@yinjunbo, yeah it works well on KITTI.
Cheers :)

**Comment by yinjunbo** [comment]
@bhardwajvijay 

Dear bhardwajvijay,

      Sorry for bothering you again, but I have one simple question to consult you.

      The flownet framework demands the resolution of input images be the  integer times of 64.However, the resolution of  KITTI benchmark is 1226x370,which should be better resized to 1216x384.

      When testing, I just use  `resize_bilinear` to the original inputs to 1216x384, and before obtaining the result I resize it to the 1226x370 again to match the lable. But the result is not ideal and error is large than the paper.

       So do you have any better solution to this problem? Or which resizing operation did you do to the images? I'm looking forward to your early reply!

All the best.
Junbo

**Comment by yinjunbo** [comment]
@bhardwajvijay

Dear bhardwajvijay,

     Sorry for bothering you again, but I have one simple question to consult you.

     The flownet framework demands the resolution of input images be the  integer times of 64.However, the resolution of  KITTI benchmark is 1226x370,which should be better resized to 1216x384.

  When testing, I just use  resize_bilinear to the original inputs to 1216x384, and before obtaining the result I resize it to the 1226x370 again to match the lable. But the result is not ideal and error is large than the paper.

   So do you have any better solution to this problem? Or which resizing operation did you do to the images? I'm looking forward to your early reply!

All the best.
Junbo

**Comment by yinjunbo** [comment]
@bhardwajvijay

Dear bhardwajvijay,

Sorry for bothering you again, but I have one simple question to consult you.

The flownet framework demands the resolution of input images be the  integer times of 64. However, the resolution of  KITTI benchmark is 1226x370,which should be better resized to 1216x384.

When testing, I just use **resize_bilinear** to the original inputs to 1216x384, and before obtaining the result I resize it to the 1226x370 again to match the lable. But the result is not ideal and error is larger than the paper.

So do you have any better solution to this problem? Or which resizing operation did you do to the images? I'm looking forward to your early reply!

All the best.
Junbo

**Comment by bhardwajvijay** [comment]
Hi @yinjunbo 

I did not resized the images. Instead I cropped the portion of input images (as well as ground truth), that would be the multiple of 64. 

So, you can do the same. Crop both input images and ground truth to the resolution 1216 x 384, and test on the same.

Regards,
Vijay

**Comment by yinjunbo** [comment]
@bhardwajvijay 

Thank you for your reply! 

According to my understanding, when cropping the width to 1216, I just need to cut 5 pixels along the edges of both left and right.

And when dealing with the height, is it to pad value 0 from 370 to 384 along  the edges to both the images and ground truth?

Sincerely,
Junbo

**Comment by bhardwajvijay** [comment]
Yes, that is right.

**Comment by yinjunbo** [comment]
OK, I'll have a try.
Thank you again!

**Comment by PoojaGhaywate** [comment]
@bhardwajvijay  Can you just help me on how to run this code on own dataset

**Comment by LiuzhuForFun** [comment]
@bhardwajvijay  Could you please solve some bugs of my codes？Or Can you give me the code you test the performance of KITTI.Thanks  a lot for your contribution of the research on optical flow.I modified the code from DFNet,but I can't get the same result you reported.My code like this:`
from __future__ import division
import cv2
import tensorflow as tf
import numpy as np
import os
import PIL.Image as pil
import png
import scipy
from ..training_schedules import LONG_SCHEDULE

from .flownet_c import FlowNetC
from ..flowlib import flow_to_image, write_flow
from scipy.misc import imread, imsave
import uuid
flags = tf.app.flags
flags.DEFINE_integer("batch_size", 1, "The size of of a sample batch")
flags.DEFINE_integer("img_height", 384, "Image height")
flags.DEFINE_integer("img_width", 1280, "Image width")
flags.DEFINE_string("dataset_dir", '', "Dataset directory")
flags.DEFINE_string("output_dir", None, "Output directory")
flags.DEFINE_string("ckpt_file", '', "checkpoint file")
FLAGS = flags.FLAGS

FLOW_SCALE = 5.0

# kitti 2012 has 194 training pairs, 195 test pairs
if 'train' in FLAGS.dataset_dir:
    NUM = 194
elif 'test' in FLAGS.dataset_dir:
    NUM = 195

def get_flow(path):
    bgr = cv2.imread(path, cv2.IMREAD_ANYCOLOR | cv2.IMREAD_ANYDEPTH)
    invalid = bgr[:, :, 0] == 0
    out_flow = (bgr[:, :, 2:0:-1].astype('f4') - 2**15) / 64.
    out_flow[invalid] = 0
    return out_flow, bgr[:, :, 0]

def compute_flow_error(gt_flow, pred_flow, mask):
    H, W, _ = gt_flow.shape
    print(pred_flow.shape)
    old_H, old_W, _ = pred_flow.shape

    # Reshape predicted flow to have same size as ground truth
    pred0 = cv2.resize(pred_flow[:,:,0], (W, H), interpolation=cv2.INTER_LINEAR) * (1.0*W/old_W)
    pred1 = cv2.resize(pred_flow[:,:,1], (W, H), interpolation=cv2.INTER_LINEAR) * (1.0*H/old_H)
    pred = np.stack((pred0, pred1), axis=-1) * FLOW_SCALE

    err = np.sqrt(np.sum(np.square(gt_flow - pred), axis=-1))
    err_valid = np.sum(err * mask) / np.sum(mask)
    return err_valid, pred

def write_flow_png(name, flow):
    H, W, _ = flow.shape
    out = np.ones((H, W, 3), dtype=np.uint64)
    out[:,:,1] = np.minimum(np.maximum(flow[:,:,1]*64.+2**15, 0), 2**16).astype(np.uint64)
    out[:,:,0] = np.minimum(np.maximum(flow[:,:,0]*64.+2**15, 0), 2**16).astype(np.uint64)
    with open(name, 'wb') as f:
        writer = png.Writer(width=W, height=H, bitdepth=16)
        im2list = out.reshape(-1, out.shape[1]*out.shape[2]).tolist()
        writer.write(f, im2list)


def pick_frame(path):
    new_files = []
  
    for i in range(NUM):
        frame1 = os.path.join(path, 'colored_0', '{:06d}'.format(i) + '_10.png')
        frame2 = os.path.join(path, 'colored_0', '{:06d}'.format(i) + '_11.png')
        new_files.append([frame1, frame2])
    return new_files

def main(_):
    new_files = pick_frame(FLAGS.dataset_dir)
    basename = os.path.basename(FLAGS.ckpt_file)

    im1_pl = tf.placeholder(dtype=tf.float32, shape=(1, FLAGS.img_height, FLAGS.img_width, 3))
    im2_pl = tf.placeholder(dtype=tf.float32, shape=(1, FLAGS.img_height, FLAGS.img_width, 3))
    model =FlowNetC()
    inputs = {
        'input_a': im1_pl,
        'input_b': im2_pl,
    }
    pred_flows = model.model(inputs, training_schedule=LONG_SCHEDULE)

    saver = tf.train.Saver([var for var in tf.all_variables()])
    config = tf.ConfigProto()
    config.gpu_options.allow_growth = True
    errs = np.zeros(NUM)

    if not FLAGS.output_dir is None and not os.path.exists(FLAGS.output_dir):
        os.makedirs(FLAGS.output_dir)

    with tf.Session(config=config) as sess:
        saver.restore(sess, FLAGS.ckpt_file)
       
        for t in range(0, len(new_files)):
            if t % 100 == 0:
                print('processing %s: %d/%d' % (basename, t, len(new_files)))
            raw_im0 = pil.open(new_files[t][0])
            raw_im1 = pil.open(new_files[t][1])
            scaled_im0 = raw_im0.resize((FLAGS.img_width, FLAGS.img_height), pil.ANTIALIAS)
            scaled_im1 = raw_im1.resize((FLAGS.img_width, FLAGS.img_height), pil.ANTIALIAS)

            
            scaled_im0 = (np.expand_dims(np.array(scaled_im0), axis=0).astype(np.float32))/255.
            scaled_im1 = (np.expand_dims(np.array(scaled_im1), axis=0).astype(np.float32))/255.
            feed_dict = {im1_pl: scaled_im0, im2_pl: scaled_im1}
            pred_flows_val = sess.run(pred_flows, feed_dict=feed_dict)
            pred_flow_val = pred_flows_val['flow']

  
            if 'train' in FLAGS.dataset_dir:
              
                #gt_flow, mask = get_flow(new_files[t][0].replace('colored_0', 'flow_noc'))
             
                gt_flow, mask = get_flow(new_files[t][0].replace('colored_0', 'flow_occ'))
                errs[t], scaled_pred = compute_flow_error(gt_flow, pred_flow_val[0,:,:,:], mask)
                unique_name = 'flow-' + str(uuid.uuid4())
                flow_img = flow_to_image(pred_flow_val[0,:,:,:])
                full_out_path = os.path.join(unique_name + '.png')
                imsave(full_out_path, flow_img)

  
            if 'test' in FLAGS.dataset_dir:
                _, scaled_pred = compute_flow_error(np.array(raw_im0)[:,:,:2], pred_flow_val[0,:,:,:], np.array(raw_im0)[:,:,0])
                png_name = os.path.join(FLAGS.output_dir, new_files[t][0].split('/')[-1])
                write_flow_png(png_name, scaled_pred)


            if not 'test' in FLAGS.dataset_dir and not FLAGS.output_dir is None:
                # flow_im = flow_to_image(scaled_pred)
                png_name = os.path.join(FLAGS.output_dir, new_files[t][0].split('/')[-1]).replace('png', 'jpg')
                # cv2.imwrite(png_name, flow_im[:,:,::-1])

        print('{:>10}'.format('(valid) endpoint error'))
        print('{:10.4f}'.format(errs.mean()))

if __name__ == '__main__':
    tf.app.run()`
